src/mqt/problemsolver/partialcompiler/qaoa.py

```diff
@@ -0,0 +1,145 @@
+                        remove_gates.append(p.name)
+                    else:
+                        remove_gates.append(False)
+                        qc_baseline.rzz(p, i, j)
+                    parameter_counter += 1
+
+            m = Parameter(f"b_{k}")
+
+            # Mixer Layer
+            qc.rx(2 * m, range(self.num_qubits))
+            qc_baseline.rx(2 * m, range(self.num_qubits))
+
+        qc.measure_all()
+        qc_baseline.measure_all()
+
+        return qc, qc_baseline, remove_gates
+
+    def compile_qc(self, baseline: bool = False, opt_level: int = 3) -> QuantumCircuit:
+        """Compiles the circuit"""
+        if baseline:
+            return transpile(self.qc_baseline, backend=self.backend, optimization_level=opt_level, seed_transpiler=42)
+        return transpile(self.qc, backend=self.backend, optimization_level=opt_level, seed_transpiler=42)
+
+    def get_to_be_removed_gate_indices(self) -> list[int]:
+        """Returns the indices of the gates to be removed"""
+        indices_parameterized_gates = []
+        for i, gate in enumerate(self.qc_compiled._data):
+            if (
+                gate.operation.name == "rz"
+                and isinstance(gate.operation.params[0], Parameter)
+                and gate.operation.params[0].name.startswith("a_")
+            ):
+                indices_parameterized_gates.append(i)
+        assert len(indices_parameterized_gates) == len(self.remove_gates)
+        indices_to_be_removed_parameterized_gates = []
+        for i in range(len(indices_parameterized_gates)):
+            if self.qc_compiled._data[indices_parameterized_gates[i]].operation.params[0].name in self.remove_gates:
+                indices_to_be_removed_parameterized_gates.append(indices_parameterized_gates[i])
+
+        assert len(set(indices_to_be_removed_parameterized_gates)) == len({elem for elem in self.remove_gates if elem})
+        return indices_to_be_removed_parameterized_gates
+
+    def remove_unnecessary_gates(self, qc: QuantumCircuit, optimize_swaps: bool = True) -> QuantumCircuit:
+        """Removes the gates to be checked from the circuit at online time"""
+        to_be_deleted_indices = []
+
+        # Iterate over all gates to be removed
+        for i in self.to_be_removed_gates_indices:
+            # Remove the gate from the ParameterTable (was needed for QCEC to work)
+            del qc._parameter_table[
+                qc._data[i].operation.params[0]
+            ]  # remove the Parameter from the ParameterTable for the specific parameter. The table just contains one entry
+            if optimize_swaps and qc._data[i - 1].operation.name == "cx" and qc._data[i - 1] == qc._data[i + 1]:
+                to_be_deleted_indices.extend([i - 1, i, i + 1])
+            else:
+                to_be_deleted_indices.append(i)
+
+        for index in sorted(to_be_deleted_indices, reverse=True):
+            del qc._data[index]
```

Comment:
This is most likely not the most efficient way to delete the respective gates from the quantum circuit.
See https://stackoverflow.com/questions/21032034/deleting-multiple-indexes-from-a-list-at-once-python.
Untested suggestion:
```suggestion
        indices = {}

        # Iterate over all gates to be removed
        for i in self.to_be_removed_gates_indices:
            # Remove the Parameter from the ParameterTable for the specific parameter.
            del qc._parameter_table[qc._data[i].operation.params[0]]
            indices.add(i)
            if optimize_swaps and qc._data[i - 1].operation.name == "cx" and qc._data[i - 1] == qc._data[i + 1]:
                indices.add(i-1)
                indices.add(i+1)

        qc._data = [v for i, v in enumerate(qc._data) if i not in indices]
```